Kvm ioctls

andreeaflorescu opened the pull request:
Added wrappers over KVM ioctls. The wrappers are split in 3 categories:
- system ioctls (as part of the `Kvm` implementation)
- VM ioctls (as part of the `VmFd` implementation) - probably I'll rename it to Vm
- Vcpu ioctls (as part of the `VcuFd` implementation) - probably I'll rename it to Vcpu

I still have to add some examples & ioctls used by CrosVM & some other tests.

andreeaflorescu committed:
sys_ioctl: added ioctl macros

This is a temporary state until we have the ioctls defines in a
separate crate.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added kvm ioctls

Added the list of kvm specific ioctls.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added wrapper over KVM capabilities

The capabilities defined in kvm-bindings are re-exported as part of the
Cap enum so they can be easily passed as function parameters.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added wrappers over KVM system ioctls

Created a Kvm structure to wrap system ioctls.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added integration tests

The integration tests are implemented on top of pytest. We cannot
used Travis (at least not out of the box) because we need a bare
metal instance that has access to /dev/kvm.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added .gitignore

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added TODO file to track missing functionality

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
Added wrappers over system ioctls used by crosvm

Added wrappers over the following ioctls:
- KVM_GET_EMULATED_CPUID
- KVM_GET_MSR_INDEX_LIST

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
docs: added the KVM ioctl used by each function

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added KVM VM ioctl wrappers

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added KVM vCPU ioctls

Added all KVM vCPU ioctls used by Firecracker.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

sameo commented on src/lib.rs:
```diff
@@ -0,0 +1,1362 @@
+    /// # Arguments
+    /// * `fd` - File descriptor to mmap from.
+    /// * `size` - Size of memory region in bytes.
+    pub fn from_fd(fd: &AsRawFd, size: usize) -> Result<KvmRunWrapper> {
```
`mmap_from_fd` or simply `mmap` would be more explicit imho.

andreeaflorescu replied:
Done.

sameo reviewed (commented):
This looks complete as far as API implementation is concerned.

andreeaflorescu committed:
split ioctls wrappers in 3 modules

Split ioctls in system, vcpu and vm ioctls so we do not have
a huge file with them all.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
adding more unit tests

- Vcpu related unit tests
- error cases for all ioctls

Signed-off-by: Andreea Florescu <fandree@amazon.com>

sameo commented on src/ioctls/vm.rs:
```diff
@@ -0,0 +1,416 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `evt` - EventFd which will be signaled. When signaling, the usual `vmexit` to userspace
```
It's not an `EventFd` anymore. I think this was the last remaining `sys_util` dependency, thanks for removing it.

andreeaflorescu replied:
True. Yap, I removed the dependencies on sys_util & memory_model. But we still need the ioctl. I will update the docs.

andreeaflorescu replied:
Done.

acatangiu reviewed (dismissed):
LGTM :+1:

alxiord commented on src/kvm_ioctls.rs:
```diff
@@ -0,0 +1,124 @@
+        let sys_fd = unsafe { open(KVM_PATH.as_ptr() as *const c_char, O_RDWR) };
+        assert!(sys_fd >= 0);
+
+        let ret = unsafe { ioctl(&File::from_raw_fd(sys_fd), KVM_GET_API_VERSION()) };
```
nit: other variables used in these tests have specific names like `vm_fd`, maybe rename the generic `ret` to `kvm_ver` or something similar?

alxiord commented on src/cap.rs:
```diff
@@ -0,0 +1,128 @@
+#[derive(Clone, Copy, Debug)]
+#[repr(u32)]
+// We are allowing docs to be missing here because this enum is a wrapper
+// over auto-generated code.
```
nit: I'd find it useful to link to the KVM API here (I was initially taken aback by the apparent lack of order and had too google a capability to find it in [kvm.h](https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/kvm.h) and see that they're listed in *numerical* order)

alxiord commented on Cargo.toml:
```diff
@@ -9,3 +9,5 @@ keywords = ["kvm"]
 
 [dependencies]
+libc = ">=0.2.39"
+kvm-bindings = ">=0.1.0"
```
nit: reverse these for alphabetical order

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+use self::kvm_ioctls::*;
+use self::sys_ioctl::*;
+
+/// Wrapper over possible Kvm Result.
```
It's not clear what "Kvm Result" means.

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+/// Wrapper over possible Kvm Result.
+pub type Result<T> = result::Result<T, io::Error>;
+
+/// Taken from Linux Kernel v4.14.13 (arch/x86/include/asm/kvm_host.h)
```
The source is super useful but the comment doesn't actually state what the constant represents.

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+    vec_with_size_in_bytes(vec_size_bytes)
+}
+
+/// A wrapper around opening and using `/dev/kvm`.
```
Isn't it more of a wrapper over the file itself?

andreeaflorescu replied:
It is a wrapper over KVM system ioctls. I guess for the high level documentation of the struct we might not need to say that it opens /dev/kvm. I will update the comment.

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+    pub fn open_with_cloexec(close_on_exec: bool) -> Result<RawFd> {
+        let open_flags = O_RDWR | if close_on_exec { O_CLOEXEC } else { 0 };
+        // Safe because we give a constant nul-terminated string and verify the result.
+        let ret = unsafe { open("/dev/kvm\0".as_ptr() as *const c_char, open_flags) };
```
Use a constant for `"/dev/kvm\0"`.

andreeaflorescu replied:
I think it reads better as is now. We are only using this constant in one place anyway and to read the code you would need to see the define for the constant as oppose to have it right there where you are doing the open.

andreeaflorescu replied:
I would rather have it as is. We aren't using that anywhere else in the code and the code reads nicely: open("/dev/kvm\0"....) versus open(dev_kvm_variable)

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+        unsafe { ioctl(self, KVM_GET_API_VERSION()) }
+    }
+
+    /// Query the availability of a particular kvm capability.
```
You can remove this line of the comment - other comments state what the function returns.

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+        unsafe { ioctl_with_val(self, KVM_CHECK_EXTENSION(), c as c_ulong) }
+    }
+
+    /// Checks if a particular `Cap` is available.
```
nit: it would be nice to add a [documentation link](https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md) for `Cap`.

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+        self.check_extension_int(c) >= 1
+    }
+
+    /// Gets the size of the mmap required to use vcpu's `kvm_run` structure.
```
```suggestion
    /// Gets the size of the memory mapping required to use the vcpu's `kvm_run` structure.
```

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+        let res = unsafe { ioctl(self, KVM_GET_VCPU_MMAP_SIZE()) };
+        if res > 0 {
+            Ok(res as usize)
+        } else {
```
This block:

```rust
if ...
else {
    Err(io::Error::last_os_error())
}
```
shows up a lot, maybe add a shorthand for it.

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+        }
+    }
+
+    /// Gets the recommended number of VCPUs per VM.
```
Please use consistent spelling for `vcpu` - either all caps, or no caps, or `vCPU`, currently it's mixed.

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+    }
+
+    /// Gets the size of the mmap required to use vcpu's `kvm_run` structure.
+    pub fn get_vcpu_mmap_size(&self) -> Result<usize> {
```
[Rustacean getters](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#gettersetter-apis) usually skip the `get_` prefix.

andreeaflorescu replied:
I would prefer having the functions here map directly to their definition from KVM API.

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+
+    /// Checks if a particular `Cap` is available.
+    ///
+    /// According to the KVM API doc, KVM_CHECK_EXTENSION returns "0 if unsupported; 1 (or some
```
```suggestion
    /// According to the KVM API doc, `KVM_CHECK_EXTENSION` returns "0 if unsupported; 1 (or some
```

alxiord commented on src/lib.rs:
```diff
@@ -0,0 +1,401 @@
+        if x > 0 {
+            x as usize
+        } else {
+            32
```
nit: use a const for this, and move the explanation to the const's doccomment.

andreeaflorescu replied:
I would rather not for the same reason in the comment about `/dev/kvm`. It think it is more of a personal preference.

alxiord commented on README.md:
````diff
@@ -1 +1,41 @@
+
+The line coverage is saved in [tests/coverage](tests/coverage). To update the
+coverage before submitting a PR, run the coverage test:
+```bash
````
nit: surround code snippets with blank lines

alxiord commented on tests/test_correctness.py:
```diff
@@ -0,0 +1,19 @@
+"""Test the correctness of the code using linters and unit tests."""
+
+import subprocess
+
```
nit: 2 blank lines

alxiord commented on tests/test_coverage.py:
```diff
@@ -0,0 +1,97 @@
+    with open(coverage_path) as f:
+        coverage = f.readline()
+    return float(coverage.strip())
+
```
nit: 2 blank lines.

alxiord commented on tests/test_coverage.py:
```diff
@@ -0,0 +1,97 @@
+
+    with open(coverage_path, "w") as f:
+        f.write(str(cov_value))
+
```
nit: 2 blank lines.

alxiord commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+/target
+**/*.rs.bk
```
What about `pychache` generated by `pytest`?

alxiord commented on TODO:
```diff
@@ -0,0 +1,8 @@
+- High Level docs with usage
```
Wouldn't issues be of more help?

andreeaflorescu replied:
Before we get the code merged, issues are kinda un-related to the code base. I can add issues after we merge this PR, I believe it makes more sense then.

alxiord commented on src/lib.rs:
```diff
@@ -319,6 +320,304 @@ pub struct VmFd {
+    ///
+    /// # Arguments
+    ///
+    /// * `evt` - EventFd which will be signaled. When signaling, the usual `vmexit` to userspace
```
Looks like it's actually a `RawFd`.

andreeaflorescu replied:
This is updated in the commit where I moved the code to its own module. I found it very tedious to update from the beginning.

alxiord commented on src/lib.rs:
```diff
@@ -319,6 +320,304 @@ pub struct VmFd {
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    pub fn get_dirty_log(&self, slot: u32, memory_size: usize) -> Result<Vec<u64>> {
+        // Compute the length of the bitmap needed for all dirty pages in one memory slot.
+        // One memory page is 4KiB (4096 bits) and KVM_GET_DIRTY_LOG returns one dirty bit for
```
```suggestion
        // One memory page is 4KiB (4096 bits) and `KVM_GET_DIRTY_LOG` returns one dirty bit for
```

alxiord commented on src/lib.rs:
```diff
@@ -319,6 +320,304 @@ pub struct VmFd {
+        // Compute the length of the bitmap needed for all dirty pages in one memory slot.
+        // One memory page is 4KiB (4096 bits) and KVM_GET_DIRTY_LOG returns one dirty bit for
+        // each page.
+        let page_size = 4 << 10;
```
Maybe use a const?

alxiord commented on src/lib.rs:
```diff
@@ -319,6 +320,304 @@ pub struct VmFd {
+    }
+}
+
+/// A wrapper around creating and using a kvm related VCPU fd
```
```suggestion
/// A wrapper around creating and using a kvm related VCPU fd.
```

alxiord replied:
Also, `KVM` is inconsistently spelled as well across the comments.

alxiord commented on src/lib.rs:
```diff
@@ -19,7 +19,7 @@ mod kvm_ioctls;
 mod cap;
```
The commit message (3f1318bdbff68f6e871e05c49698b3725fb7af20)says `KVM Vcpus` instead of (I think) `KVM ioctls`.

alxiord commented on src/lib.rs:
```diff
@@ -712,6 +1071,26 @@ mod tests {
         }
     }
 
+    // Helper function for mmap an anonymous memory of `size`.
```
```suggestion
    // Helper function for memory mapping `size` bytes of anonymous memory.
```

alxiord commented on src/lib.rs:
```diff
@@ -712,6 +1071,26 @@ mod tests {
+            libc::mmap(
+                null_mut(),
+                size,
+                libc::PROT_READ | libc::PROT_WRITE,
```
Maybe `use` all these constants fro `libc`?

alxiord commented on src/lib.rs:
```diff
@@ -886,4 +1265,98 @@ mod tests {
+
+        let kvm = Kvm::new().unwrap();
+        let vm = kvm.create_vm().unwrap();
+        // This example based on https://lwn.net/Articles/658511/
```
nit:
```suggestion
        // This example is based on https://lwn.net/Articles/658511/ .
```

alxiord commented on src/lib.rs:
```diff
@@ -886,4 +1265,98 @@ mod tests {
+            0xec, /* in %dx, %al */
+            0xc6, 0x06, 0x00, 0x80, 0x00, /* movl $0, (0x8000) */
+            0x8a, 0x16, 0x00, 0x80, /* movl (0x8000), %dl */
+            0xc6, 0x06, 0x00, 0x20, 0x00, /* movl $0, (0x2000) */
```
The comments in the [Firecracker codebase](https://github.com/firecracker-microvm/firecracker/blob/master/kvm/src/lib.rs#L1411) would help.

alxiord commented on src/ioctls/system.rs:
```diff
@@ -334,4 +334,31 @@ mod tests {
+        };
+
+        assert_eq!(get_raw_errno(faulty_kvm.get_vcpu_mmap_size()), badf_errno);
+
```
nit: I don't think blank lines particularly improve readability here

alxiord commented on src/ioctls/vcpu.rs:
```diff
@@ -417,6 +422,144 @@ mod tests {
+    #[test]
+    fn lapic_test() {
+        use std::io::Cursor;
+        //we might get read of byteorder if we replace 5h3 mem::transmute with something safer
```
What is `5h3`?

andreeaflorescu replied:
3 characters from one of my passwords, but I won't tell you which one! I'll remove it.

alxiord commented on src/ioctls/vcpu.rs:
```diff
@@ -417,6 +422,144 @@ mod tests {
+        use std::io::Cursor;
+        //we might get read of byteorder if we replace 5h3 mem::transmute with something safer
+        use self::byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+        //as per https://github.com/torvalds/linux/arch/x86/kvm/lapic.c
```
Add a blank space and punctuation.

alxiord replied:
This applies to following comments as well.

andreeaflorescu committed:
Updated documentation

Enhanced the documentation for various public structures and constants.
Mostly added links and added more details around what are the structures
or constants used for.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
updated .gitignore to ignore pycache

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu commented:
@aghecenco I have added 2 more commits. I wasn't able to fix all the nits, but I will do another pass before we actually publish the kvm-ioctls crate to crates.io.

I'll make sure that the documentation is up to date. If it is okay with you I would say we merge it as it is right now and add smaller PRs to fix the other nits.

alxiord reviewed (approved):
Remaining style nits to be addressed in a future PR (this one already has 15 commits :slightly_smiling_face:)

sameo reviewed (approved):
LGTM

sboeuf commented:
It would be nice to have input from @zachreizner before we decide to merge this crate as the point being to make sure this can be properly reused by `crosvm`.

sameo commented:
@dgreid gave the green light on the rust-vmm mailing list:
http://lists.opendev.org/pipermail/rust-vmm/2019-March/000180.html

This one is good to go, I'm merging it.
